pkg/ipam-controller/controllers/node/node.go

```diff
@@ -15,31 +15,120 @@ package controllers
 
 import (
 	"context"
+	"errors"
+	"reflect"
+	"time"
 
+	apiErrors "k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/workqueue"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
 	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
 
 	corev1 "k8s.io/api/core/v1"
+
+	"github.com/Mellanox/nvidia-k8s-ipam/pkg/ipam-controller/allocator"
+	"github.com/Mellanox/nvidia-k8s-ipam/pkg/ipam-controller/selector"
+	"github.com/Mellanox/nvidia-k8s-ipam/pkg/pool"
 )
 
 // NodeReconciler reconciles Node objects
 type NodeReconciler struct {
+	Allocator *allocator.Allocator
+	Selector  *selector.Selector
+
+	ConfigEventCh chan event.GenericEvent
 	client.Client
 	Scheme *runtime.Scheme
 }
 
 // Reconcile contains logic to sync Node objects
-func (r *NodeReconciler) Reconcile(ctx context.Context, _ ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+func (r *NodeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	reqLog := log.FromContext(ctx)
+	if !r.Allocator.IsConfigured() {
+		reqLog.V(1).Info("allocator is not yet configured, requeue")
+		return ctrl.Result{RequeueAfter: time.Second, Requeue: true}, nil
+	}
+
+	node := &corev1.Node{}
+	err := r.Client.Get(ctx, req.NamespacedName, node)
+	if err != nil {
+		if apiErrors.IsNotFound(err) {
+			reqLog.Info("node object removed, deallocate ranges")
+			r.Allocator.Deallocate(ctx, req.Name)
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	if !r.Selector.Match(node) {
+		reqLog.Info("node doesn't match selector, ensure range is not allocated")
+		r.Allocator.Deallocate(ctx, node.Name)
+		return r.cleanAnnotation(ctx, node)
+	}
+
+	var existingNodeAlloc map[string]*pool.IPPool
+	nodeAllocManager, err := pool.NewManagerImpl(node)
+	if err == nil {
+		existingNodeAlloc = nodeAllocManager.GetPools()
+	}
+
+	expectedAlloc, err := r.Allocator.Allocate(ctx, node.Name)
+	if err != nil {
+		if errors.Is(allocator.ErrNoFreeRanges, err) {
+			return r.cleanAnnotation(ctx, node)
+		}
+		return ctrl.Result{}, err
+	}
+	if reflect.DeepEqual(existingNodeAlloc, expectedAlloc) {
+		reqLog.Info("node ranges are up-to-date")
+		return ctrl.Result{}, nil
+	}
+
+	if err := pool.SetPools(node, expectedAlloc); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if err := r.Client.Update(ctx, node); err != nil {
```

Comment:
should we use patch here ? to reduce chance of error ?

Reply:
cool thx, in the past we got folks complaining on errors in controller, if we can avoid them its better.
(will save time in the long term ...)